### PR TITLE
Add description formatting utilities

### LINKED ISSUE #259: Add support for a format specifier in event description strings to format a payload value when displayed

### PR DESCRIPTION
**Pull request by jonlamb-gh**
Adds a trait providing description string payload formatting.

Closes #259

### CODE REVIEW

**Comment by nathanaschbacher** [comment]
What does this look like to use from both Rust and C? Does this also include the "glue" to the code to do this when the manifest is consumed to decorate the raw data with more context?

**Comment by jonlamb-gh** [comment]
> What does this look like to use from both Rust and C? Does this also include the "glue" to the code to do this when the manifest is consumed to decorate the raw data with more context?

Check out the diff in this PR, I added a formatted description to both the C and the Rust examples. This only adds the capability to format a dynamic string with a payload value, was thinking it'd be glue'd together in #262 as part of `log`.

**Comment by pittma** [comment]
> > What does this look like to use from both Rust and C? Does this also include the "glue" to the code to do this when the manifest is consumed to decorate the raw data with more context?
> 
> Check out the diff in this PR, I added a formatted description to both the C and the Rust examples. This only adds the capability to format a dynamic string with a payload value, was thinking it'd be glue'd together in #262 as part of `log`.

This is what I'm thinking as well, and it's why I left the `log` PR as a draft. The only additional thought I had about this specifically is that I wonder if it would be useful to really lean into the logging disposition and add log level-specific definitions that map to severity levels. Something like you get with Rust logging with `log::info!(…)`.

**Comment by nathanaschbacher** [comment]
> > > What does this look like to use from both Rust and C? Does this also include the "glue" to the code to do this when the manifest is consumed to decorate the raw data with more context?
> > 
> > 
> > Check out the diff in this PR, I added a formatted description to both the C and the Rust examples. This only adds the capability to format a dynamic string with a payload value, was thinking it'd be glue'd together in #262 as part of `log`.
> 
> This is what I'm thinking as well, and it's why I left the `log` PR as a draft. The only additional thought I had about this specifically is that I wonder if it would be useful to really lean into the logging disposition and add log level-specific definitions that map to severity levels. Something like you get with Rust logging with `log::info!(…)`.

I like that last bit, and I could totally see that sort of functionality as part of the "Log friendly API effort" here #264

**Comment by nathanaschbacher** [comment]
I think what needs to happen to merge this in as a first pass is to update the `export` capability to glue together the SUT event description w/ the payload in the output.

**Comment by pittma** [comment]
> I think what needs to happen to merge this in as a first pass is to update the `export` capability to glue together the SUT event description w/ the payload in the output.

I can take a stab at that once the `log` design settles.

**Comment by pittma** [comment]
> I can take a stab at that once the `log` design settles.

https://github.com/auxoncorp/modality-probe/pull/261/commits/f8db997a05fa78d57ca802e8a83ffff91292ed6d interpolates payloads into descriptions if the description contains `{}`:

```
        PRODUCER_MEASUREMENT_SAMPLED_PRODUCER_PROBE_0_3 [
                label        = "PRODUCER_MEASUREMENT_SAMPLED"
                
                message      = "Measurement producer sampled a value for transmission: 2"
                
                file         = "rust-example/src/main.rs" 
                probe        = "PRODUCER_PROBE"
                tags         = "producer;measurement sample"
                
                payload      = 2                      
                                                 
                raw_event_id = 2
                raw_probe_id = 837318897  
        ];
```

File: modality-probe-cli/src/visualize/graph.rs [hunk 1]
```diff
@@ -219,13 +219,26 @@ fn graph_to_tree<'a>(
...
-                meta::parsed_payload(emeta.type_hint.as_ref().map(|s| s.as_ref()), node.payload)
-                    .unwrap_or(None);
+            let payload = node.payload.and_then(|pl| {
+                emeta.type_hint.as_ref().and_then(|th| {
+                    meta::parsed_payload(Some(th.as_ref()), Some(pl))
+                        .ok()
+                        .flatten()
+                })
+            });
+            let has_log_str = emeta.description.contains("{}");
```
Comment (jonlamb-gh) [inline modality-probe-cli/src/visualize/graph.rs:229]: The `DescriptionFormat` trait has a `contains_formatting()` method for this.
Should make it easier for us to expand later on for extra formatting, like `{:X}` for hex.

File: modality-probe-cli/src/visualize/graph.rs [hunk 2]
```diff
@@ -219,13 +219,26 @@ fn graph_to_tree<'a>(
...
+            let log_str = if has_log_str && payload.is_some() {
+                Some(emeta.description.replace("{}", payload.as_ref().unwrap()))
```
Comment (jonlamb-gh) [inline modality-probe-cli/src/visualize/graph.rs:231]: Maybe this should be using the `format_payload()` method from the `DescriptionFormat` trait?

Comment (pittma) [reply modality-probe-cli/src/visualize/graph.rs:231]: Yep! just made this change. Pushing shortly.

**Pull request merged**